// molstar/lib/commonjs/mol-repr/volume/slice.js
"use strict";
/**
 * Copyright (c) 2020-2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SliceRepresentationProvider = exports.SliceParams = void 0;
exports.getSliceParams = getSliceParams;
exports.createImage = createImage;
exports.SliceVisual = SliceVisual;
exports.SliceRepresentation = SliceRepresentation;
const param_definition_1 = require("../../mol-util/param-definition.js");
const image_1 = require("../../mol-geo/geometry/image/image.js");
const volume_1 = require("../../mol-model/volume.js");
const representation_1 = require("./representation.js");
const location_iterator_1 = require("../../mol-geo/util/location-iterator.js");
const picking_1 = require("../../mol-geo/geometry/picking.js");
const loci_1 = require("../../mol-model/loci.js");
const util_1 = require("../../mol-geo/util.js");
const color_1 = require("../../mol-util/color/index.js");
const color_2 = require("../../mol-theme/color.js");
const number_packing_1 = require("../../mol-util/number-packing.js");
const util_2 = require("./util.js");
const vec3_1 = require("../../mol-math/linear-algebra/3d/vec3.js");
const quat_1 = require("../../mol-math/linear-algebra/3d/quat.js");
const misc_1 = require("../../mol-math/misc.js");
const mat4_1 = require("../../mol-math/linear-algebra/3d/mat4.js");
const interpolate_1 = require("../../mol-math/interpolate.js");
const type_helpers_1 = require("../../mol-util/type-helpers.js");
const ordered_set_1 = require("../../mol-data/int/ordered-set.js");
const interval_1 = require("../../mol-data/int/interval.js");
const visual_1 = require("./visual.js");
const hash_functions_1 = require("../../mol-data/util/hash-functions.js");
exports.SliceParams = {
    ...image_1.Image.Params,
    quality: { ...image_1.Image.Params.quality, isEssential: false },
    dimension: param_definition_1.ParamDefinition.MappedStatic('x', {
        x: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 0, step: 1 }, { immediateUpdate: true }),
        y: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 0, step: 1 }, { immediateUpdate: true }),
        z: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 0, step: 1 }, { immediateUpdate: true }),
        relativeX: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 1, step: 0.01 }, { immediateUpdate: true }),
        relativeY: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 1, step: 0.01 }, { immediateUpdate: true }),
        relativeZ: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 1, step: 0.01 }, { immediateUpdate: true }),
    }, { isEssential: true, hideIf: p => p.mode !== 'grid', description: 'Slice position in grid coordinates.' }),
    isoValue: volume_1.Volume.IsoValueParam,
    mode: param_definition_1.ParamDefinition.Select('grid', param_definition_1.ParamDefinition.arrayToOptions(['grid', 'frame', 'plane']), { description: 'Grid: slice through the volume along the grid axes in integer steps. Frame: slice through the volume along arbitrary axes in any step size. Plane: an arbitrary plane defined by point and normal.' }),
    offset: param_definition_1.ParamDefinition.Numeric(0, { min: -1, max: 1, step: 0.01 }, { isEssential: true, immediateUpdate: true, hideIf: p => p.mode !== 'frame', description: 'Relative offset from center.' }),
    axis: param_definition_1.ParamDefinition.Select('a', param_definition_1.ParamDefinition.arrayToOptions(['a', 'b', 'c']), { isEssential: true, hideIf: p => p.mode !== 'frame', description: 'Axis of the frame.' }),
    rotation: param_definition_1.ParamDefinition.Group({
        axis: param_definition_1.ParamDefinition.Vec3(vec3_1.Vec3.create(1, 0, 0), {}, { description: 'Axis of rotation' }),
        angle: param_definition_1.ParamDefinition.Numeric(0, { min: -180, max: 180, step: 1 }, { immediateUpdate: true, description: 'Axis rotation angle in Degrees' }),
    }, { isExpanded: true, hideIf: p => p.mode !== 'frame' }),
    plane: param_definition_1.ParamDefinition.Group({
        point: param_definition_1.ParamDefinition.Vec3(vec3_1.Vec3.create(0, 0, 0), {}, { description: 'Plane point' }),
        normal: param_definition_1.ParamDefinition.Vec3(vec3_1.Vec3.create(1, 0, 0), {}, { description: 'Plane normal' }),
    }, { isExpanded: true, hideIf: p => p.mode !== 'plane' }),
};
function getSliceParams(ctx, volume) {
    const periodicRange = volume_1.Volume.getPeriodicRange(volume);
    if (periodicRange) {
        const { min, max } = periodicRange;
        volume = (0, util_2.createPeriodicVolume)(volume, min, max);
    }
    const p = param_definition_1.ParamDefinition.clone(exports.SliceParams);
    const dim = volume.grid.cells.space.dimensions;
    p.dimension = param_definition_1.ParamDefinition.MappedStatic('x', {
        x: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: dim[0] - 1, step: 1 }, { immediateUpdate: true }),
        y: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: dim[1] - 1, step: 1 }, { immediateUpdate: true }),
        z: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: dim[2] - 1, step: 1 }, { immediateUpdate: true }),
        relativeX: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 1, step: 0.01 }, { immediateUpdate: true }),
        relativeY: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 1, step: 0.01 }, { immediateUpdate: true }),
        relativeZ: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 1, step: 0.01 }, { immediateUpdate: true }),
    }, { isEssential: true, hideIf: p => p.mode !== 'grid' });
    p.isoValue = volume_1.Volume.createIsoValueParam(volume_1.Volume.IsoValue.absolute(volume.grid.stats.min), volume.grid.stats);
    return p;
}
async function createImage(ctx, volume, key, theme, props, image) {
    const periodicRange = volume_1.Volume.getPeriodicRange(volume);
    if (periodicRange) {
        const { min, max } = periodicRange;
        volume = (0, util_2.createPeriodicVolume)(volume, min, max);
    }
    switch (props.mode) {
        case 'frame':
            return createFrameImage(ctx, volume, key, theme, props, image);
        case 'grid':
            return createGridImage(ctx, volume, key, theme, props, image);
        case 'plane':
            return createPlaneImage(ctx, volume, key, theme, props, image);
        default:
            (0, type_helpers_1.assertUnreachable)(props.mode);
    }
}
//
function getFrame(volume, props) {
    const { axis, rotation, mode } = props;
    const gridToCartn = volume_1.Grid.getGridToCartesianTransform(volume.grid);
    const cartnToGrid = mat4_1.Mat4.invert((0, mat4_1.Mat4)(), gridToCartn);
    const [nx, ny, nz] = volume.grid.cells.space.dimensions;
    const a = nx;
    const b = ny;
    const c = nz;
    const dirA = vec3_1.Vec3.create(a, 0, 0);
    const dirB = vec3_1.Vec3.create(0, b, 0);
    const dirC = vec3_1.Vec3.create(0, 0, c);
    const resolution = Math.max(a, b, c) / Math.max(nx, ny, nz);
    const min = vec3_1.Vec3.create(0, 0, 0);
    const max = vec3_1.Vec3.create(a, b, c);
    vec3_1.Vec3.transformMat4(min, min, gridToCartn);
    vec3_1.Vec3.transformMat4(max, max, gridToCartn);
    const center = vec3_1.Vec3.center((0, vec3_1.Vec3)(), max, min);
    const size = (0, vec3_1.Vec3)();
    const major = (0, vec3_1.Vec3)();
    const minor = (0, vec3_1.Vec3)();
    const normal = (0, vec3_1.Vec3)();
    if (axis === 'c') {
        vec3_1.Vec3.set(size, a, b, c);
        vec3_1.Vec3.copy(major, dirA);
        vec3_1.Vec3.copy(minor, dirB);
        vec3_1.Vec3.copy(normal, dirC);
    }
    else if (axis === 'b') {
        vec3_1.Vec3.set(size, a, c, b);
        vec3_1.Vec3.copy(major, dirA);
        vec3_1.Vec3.copy(normal, dirB);
        vec3_1.Vec3.copy(minor, dirC);
    }
    else {
        vec3_1.Vec3.set(size, b, c, a);
        vec3_1.Vec3.copy(normal, dirA);
        vec3_1.Vec3.copy(major, dirB);
        vec3_1.Vec3.copy(minor, dirC);
    }
    if (rotation.angle !== 0) {
        const ra = (0, vec3_1.Vec3)();
        vec3_1.Vec3.scaleAndAdd(ra, ra, dirA, rotation.axis[0]);
        vec3_1.Vec3.scaleAndAdd(ra, ra, dirB, rotation.axis[1]);
        vec3_1.Vec3.scaleAndAdd(ra, ra, dirC, rotation.axis[2]);
        vec3_1.Vec3.normalize(ra, ra);
        const rm = mat4_1.Mat4.fromRotation((0, mat4_1.Mat4)(), (0, misc_1.degToRad)(rotation.angle), ra);
        vec3_1.Vec3.transformDirection(major, major, rm);
        vec3_1.Vec3.transformDirection(minor, minor, rm);
        vec3_1.Vec3.transformDirection(normal, normal, rm);
    }
    if (mode === 'frame') {
        const r = vec3_1.Vec3.distance(min, max);
        const s = vec3_1.Vec3.distance(min, max);
        vec3_1.Vec3.set(size, s, s, r);
    }
    vec3_1.Vec3.transformDirection(major, major, gridToCartn);
    vec3_1.Vec3.transformDirection(minor, minor, gridToCartn);
    vec3_1.Vec3.transformDirection(normal, normal, gridToCartn);
    const trim = {
        type: 3,
        center: vec3_1.Vec3.create(a / 2, b / 2, c / 2),
        scale: vec3_1.Vec3.create(a, b, c),
        rotation: quat_1.Quat.identity(),
        transform: cartnToGrid,
    };
    return { size, major, minor, normal, center, trim, resolution };
}
function getSampledImage(volume, theme, info, isoValue, trim, image) {
    const { m, width, height } = info;
    const { cells: { space }, stats } = volume.grid;
    const { min, max } = stats;
    const isUniform = theme.color.granularity === 'uniform';
    const color = 'color' in theme.color && theme.color.color
        ? theme.color.color
        : () => (0, color_1.Color)(0xffffff);
    const v = (0, vec3_1.Vec3)();
    const gridToCartn = volume_1.Grid.getGridToCartesianTransform(volume.grid);
    const cartnToGrid = mat4_1.Mat4.invert((0, mat4_1.Mat4)(), gridToCartn);
    const [mx, my, mz] = space.dimensions;
    const imageArray = new Uint8Array(width * height * 4);
    const groupArray = new Uint8Array(width * height * 4);
    const valueArray = new Float32Array(width * height);
    const mapping = {
        instance: new Int16Array(width * height),
        cell: new Uint32Array(width * height),
        index: new Map(),
    };
    const gridCoords = (0, vec3_1.Vec3)();
    const pl = (0, location_iterator_1.PositionLocation)((0, vec3_1.Vec3)(), (0, vec3_1.Vec3)());
    const getTrilinearlyInterpolated = volume_1.Grid.makeGetTrilinearlyInterpolated(volume.grid, 'none');
    const periodicMapping = volume_1.Volume.getPeriodicMapping(volume);
    let i = 0;
    for (let ih = 0; ih < height; ++ih) {
        for (let iw = 0; iw < width; ++iw) {
            const i4 = i * 4;
            const y = ((0, interpolate_1.clamp)(iw + 0.5, 0, width - 1) / width) - 0.5;
            const x = ((0, interpolate_1.clamp)(ih + 0.5, 0, height - 1) / height) - 0.5;
            vec3_1.Vec3.set(v, x, -y, 0);
            vec3_1.Vec3.transformMat4(v, v, m);
            vec3_1.Vec3.copy(gridCoords, v);
            vec3_1.Vec3.transformMat4(gridCoords, gridCoords, cartnToGrid);
            const ix = Math.trunc(gridCoords[0]);
            const iy = Math.trunc(gridCoords[1]);
            const iz = Math.trunc(gridCoords[2]);
            vec3_1.Vec3.copy(pl.position, v);
            const c = color(pl, false);
            color_1.Color.toArray(c, imageArray, i4);
            const val = (0, interpolate_1.normalize)(getTrilinearlyInterpolated(v), min, max);
            if (isUniform) {
                imageArray[i4] *= val;
                imageArray[i4 + 1] *= val;
                imageArray[i4 + 2] *= val;
            }
            valueArray[i] = val;
            if (ix >= 0 && ix < mx && iy >= 0 && iy < my && iz >= 0 && iz < mz) {
                (0, number_packing_1.packIntToRGBArray)(i, groupArray, i4);
                if (periodicMapping) {
                    const pm = periodicMapping.get(ix, iy, iz);
                    if (pm) {
                        mapping.instance[i] = pm.instance;
                        mapping.cell[i] = pm.cell;
                        const key = (0, hash_functions_1.cantorPairing)(pm.instance, pm.cell);
                        if (mapping.index.has(key)) {
                            mapping.index.get(key).push(i);
                        }
                        else {
                            mapping.index.set(key, [i]);
                        }
                    }
                    else {
                        imageArray[i4] = 0;
                        imageArray[i4 + 1] = 0;
                        imageArray[i4 + 2] = 0;
                        imageArray[i4 + 3] = 0;
                        valueArray[i] = NaN;
                    }
                }
                else {
                    const o = space.dataOffset(ix, iy, iz);
                    mapping.instance[i] = -1;
                    mapping.cell[i] = o;
                    if (mapping.index.has(o)) {
                        mapping.index.get(o).push(i);
                    }
                    else {
                        mapping.index.set(o, [i]);
                    }
                }
            }
            i += 1;
        }
    }
    const imageTexture = { width, height, array: imageArray, flipY: true };
    const groupTexture = { width, height, array: groupArray, flipY: true };
    const valueTexture = { width, height, array: valueArray, flipY: true };
    const corners = new Float32Array([
        -0.5, 0.5, 0,
        0.5, 0.5, 0,
        -0.5, -0.5, 0,
        0.5, -0.5, 0
    ]);
    (0, util_1.transformPositionArray)(m, corners, 0, 4);
    const isoLevel = (0, interpolate_1.clamp)((0, interpolate_1.normalize)(volume_1.Volume.IsoValue.toAbsolute(isoValue, stats).absoluteValue, min, max), 0, 1);
    const im = image_1.Image.create(imageTexture, corners, groupTexture, valueTexture, trim, isoLevel, image);
    im.setBoundingSphere(volume_1.Volume.isPeriodic(volume) ? volume_1.Volume.getBoundingSphere(volume) : volume_1.Grid.getBoundingSphere(volume.grid));
    im.meta.mapping = mapping;
    return im;
}
async function createFrameImage(ctx, volume, key, theme, props, image) {
    const { offset, isoValue } = props;
    const { size, major, minor, normal, center, trim, resolution } = getFrame(volume, props);
    const scaleFactor = 1 / resolution;
    const scale = vec3_1.Vec3.create(size[0], size[1], 1);
    const offsetDir = vec3_1.Vec3.setMagnitude((0, vec3_1.Vec3)(), normal, size[2] / 2);
    const width = Math.floor(size[1] * scaleFactor);
    const height = Math.floor(size[0] * scaleFactor);
    const m = mat4_1.Mat4.identity();
    const v = (0, vec3_1.Vec3)();
    const anchor = (0, vec3_1.Vec3)();
    vec3_1.Vec3.add(v, center, major);
    mat4_1.Mat4.targetTo(m, center, v, minor);
    vec3_1.Vec3.scaleAndAdd(anchor, center, offsetDir, offset);
    mat4_1.Mat4.setTranslation(m, anchor);
    mat4_1.Mat4.mul(m, m, mat4_1.Mat4.rotY90);
    mat4_1.Mat4.scale(m, m, scale);
    const info = { m, width, height };
    return getSampledImage(volume, theme, info, isoValue, trim, image);
}
async function createPlaneImage(ctx, volume, key, theme, props, image) {
    const { plane: { point, normal }, isoValue } = props;
    const nn = vec3_1.Vec3.normalize((0, vec3_1.Vec3)(), normal);
    const gridToCartn = volume_1.Grid.getGridToCartesianTransform(volume.grid);
    const cartnToGrid = mat4_1.Mat4.invert((0, mat4_1.Mat4)(), gridToCartn);
    const [mx, my, mz] = volume.grid.cells.space.dimensions;
    const a = mx;
    const b = my;
    const c = mz;
    const resolution = Math.max(a, b, c) / Math.max(mx, my, mz);
    const scaleFactor = 1 / resolution;
    const s = vec3_1.Vec3.distance(vec3_1.Vec3.create(0, 0, 0), vec3_1.Vec3.create(a, b, c));
    // Get center on plane
    const center = vec3_1.Vec3.create(a / 2, b / 2, c / 2);
    vec3_1.Vec3.transformMat4(center, center, gridToCartn);
    const toCenter = vec3_1.Vec3.sub((0, vec3_1.Vec3)(), center, point);
    const distToPlane = vec3_1.Vec3.dot(toCenter, nn);
    const centerOnPlane = vec3_1.Vec3.scaleAndAdd((0, vec3_1.Vec3)(), center, nn, -distToPlane);
    // Center image on volume (within plane)
    const m = mat4_1.Mat4.fromPlane((0, mat4_1.Mat4)(), nn, centerOnPlane);
    const size = vec3_1.Vec3.set((0, vec3_1.Vec3)(), s, s, s);
    mat4_1.Mat4.scale(m, m, size);
    const width = Math.floor(s * scaleFactor);
    const height = Math.floor(s * scaleFactor);
    const trim = {
        type: 3,
        center: vec3_1.Vec3.create(a / 2, b / 2, c / 2),
        scale: vec3_1.Vec3.create(a, b, c),
        rotation: quat_1.Quat.identity(),
        transform: cartnToGrid,
    };
    const info = { m, width, height };
    return getSampledImage(volume, theme, info, isoValue, trim, image);
}
async function createGridImage(ctx, volume, key, theme, props, image) {
    const { dimension: { name: dim }, isoValue } = props;
    const { cells: { space, data }, stats } = volume.grid;
    const { min, max } = stats;
    const isUniform = theme.color.granularity === 'uniform';
    const color = 'color' in theme.color && theme.color.color
        ? theme.color.color
        : () => (0, color_1.Color)(0xffffff);
    const { width, height, x, y, z, x0, y0, z0, nx, ny, nz } = getSliceInfo(volume.grid, props);
    const corners = new Float32Array(dim === 'x' || dim === 'relativeX' ? [x, 0, 0, x, y, 0, x, 0, z, x, y, z] :
        dim === 'y' || dim === 'relativeY' ? [0, y, 0, x, y, 0, 0, y, z, x, y, z] :
            [0, 0, z, 0, y, z, x, 0, z, x, y, z]);
    const imageArray = new Uint8Array(width * height * 4);
    const groupArray = new Uint8Array(width * height * 4);
    const valueArray = new Float32Array(width * height);
    const mapping = {
        instance: new Int16Array(width * height),
        cell: new Uint32Array(width * height),
        index: new Map(),
    };
    const gridToCartn = volume_1.Grid.getGridToCartesianTransform(volume.grid);
    const l = (0, location_iterator_1.PositionLocation)((0, vec3_1.Vec3)(), (0, vec3_1.Vec3)());
    const periodicMapping = volume_1.Volume.getPeriodicMapping(volume);
    let i = 0;
    for (let iy = y0; iy < ny; ++iy) {
        for (let ix = x0; ix < nx; ++ix) {
            for (let iz = z0; iz < nz; ++iz) {
                const i4 = i * 4;
                vec3_1.Vec3.set(l.position, ix, iy, iz);
                vec3_1.Vec3.transformMat4(l.position, l.position, gridToCartn);
                color_1.Color.toArray(color(l, false), imageArray, i4);
                const val = (0, interpolate_1.normalize)(space.get(data, ix, iy, iz), min, max);
                if (isUniform) {
                    imageArray[i4] *= val;
                    imageArray[i4 + 1] *= val;
                    imageArray[i4 + 2] *= val;
                }
                valueArray[i] = val;
                (0, number_packing_1.packIntToRGBArray)(i, groupArray, i4);
                if (periodicMapping) {
                    const pm = periodicMapping.get(ix, iy, iz);
                    if (pm) {
                        mapping.instance[i] = pm.instance;
                        mapping.cell[i] = pm.cell;
                        mapping.index.set((0, hash_functions_1.cantorPairing)(pm.instance, pm.cell), [i]);
                    }
                    else {
                        imageArray[i4] = 0;
                        imageArray[i4 + 1] = 0;
                        imageArray[i4 + 2] = 0;
                        imageArray[i4 + 3] = 0;
                        valueArray[i] = NaN;
                    }
                }
                else {
                    const o = space.dataOffset(ix, iy, iz);
                    mapping.instance[i] = -1;
                    mapping.cell[i] = o;
                    mapping.index.set(o, [i]);
                }
                i += 1;
            }
        }
    }
    const imageTexture = { width, height, array: imageArray, flipY: true };
    const groupTexture = { width, height, array: groupArray, flipY: true };
    const valueTexture = { width, height, array: valueArray, flipY: true };
    const transform = volume_1.Grid.getGridToCartesianTransform(volume.grid);
    (0, util_1.transformPositionArray)(transform, corners, 0, 4);
    const trim = image_1.Image.createEmptyTrim();
    const isoLevel = (0, interpolate_1.clamp)((0, interpolate_1.normalize)(volume_1.Volume.IsoValue.toAbsolute(isoValue, stats).absoluteValue, min, max), 0, 1);
    const im = image_1.Image.create(imageTexture, corners, groupTexture, valueTexture, trim, isoLevel, image);
    im.setBoundingSphere(volume_1.Volume.isPeriodic(volume) ? volume_1.Volume.getBoundingSphere(volume) : volume_1.Grid.getBoundingSphere(volume.grid));
    im.meta.mapping = mapping;
    return im;
}
//
function getRelativeIndex(dim, index) {
    return (0, interpolate_1.clamp)(Math.round((dim - 1) * index), 0, dim - 1);
}
function getSliceInfo(grid, props) {
    const { dimension: { name: dim, params: index } } = props;
    const { space } = grid.cells;
    let width, height;
    let x, y, z;
    let x0 = 0, y0 = 0, z0 = 0;
    let [nx, ny, nz] = space.dimensions;
    if (dim === 'x') {
        x = index, y = ny, z = nz;
        width = nz, height = ny;
        x0 = x, nx = x0 + 1;
    }
    else if (dim === 'y') {
        x = nx, y = index, z = nz;
        width = nz, height = nx;
        y0 = y, ny = y0 + 1;
    }
    else if (dim === 'z') {
        x = nx, y = ny, z = index;
        width = nx, height = ny;
        z0 = z, nz = z0 + 1;
    }
    else if (dim === 'relativeX') {
        x = getRelativeIndex(nx, index);
        y = ny;
        z = nz;
        width = nz, height = ny;
        x0 = x, nx = x0 + 1;
    }
    else if (dim === 'relativeY') {
        x = nx;
        y = getRelativeIndex(ny, index);
        z = nz;
        width = nz, height = nx;
        y0 = y, ny = y0 + 1;
    }
    else /* if (dim === 'relativeZ') */ {
        x = nx;
        y = ny;
        z = getRelativeIndex(nz, index);
        width = nx, height = ny;
        z0 = z, nz = z0 + 1;
    }
    return {
        width, height,
        x, y, z,
        x0, y0, z0,
        nx, ny, nz
    };
}
function getObjectLoci(volume, instances, props) {
    // TODO: get exact groups
    return volume_1.Volume.Loci(volume, instances);
}
function getLoci(volume, props) {
    const instances = interval_1.Interval.ofLength(volume.instances.length);
    return getObjectLoci(volume, instances, props);
}
function getSliceLoci(pickingId, volume, _key, props, id, image) {
    let { objectId, groupId, instanceId } = pickingId;
    if (id === objectId) {
        const mapping = image.meta.mapping;
        if (mapping) {
            const instanceIndex = mapping.instance[groupId];
            if (instanceIndex >= 0)
                instanceId = instanceIndex;
            groupId = mapping.cell[groupId];
        }
        const granularity = volume_1.Volume.PickingGranularity.get(volume);
        const instances = ordered_set_1.OrderedSet.ofSingleton(instanceId);
        if (granularity === 'volume') {
            return volume_1.Volume.Loci(volume, instances);
        }
        if (granularity === 'object' || groupId === picking_1.PickingId.Null) {
            return getObjectLoci(volume, instances, props);
        }
        else {
            const indices = interval_1.Interval.ofSingleton(groupId);
            return volume_1.Volume.Cell.Loci(volume, [{ indices, instances }]);
        }
    }
    return loci_1.EmptyLoci;
}
function eachSlice(loci, volume, key, props, apply, image) {
    const mapping = image.meta.mapping;
    if (mapping) {
        const groupCount = mapping.cell.length;
        const cellCount = volume.grid.cells.data.length;
        const isPeriodic = volume_1.Volume.isPeriodic(volume);
        const getIndices = isPeriodic
            ? (instanceIndex, groupIndex) => {
                return mapping.index.get((0, hash_functions_1.cantorPairing)(instanceIndex, groupIndex));
            }
            : (instanceIndex, groupIndex) => {
                const indices = mapping.index.get(groupIndex);
                return indices !== undefined ? indices.map(idx => idx + instanceIndex * groupCount) : undefined;
            };
        return (0, util_2.eachVolumeLoci)(loci, volume, undefined, (interval) => {
            let changed = false;
            for (let i = interval_1.Interval.start(interval), il = interval_1.Interval.end(interval); i < il; ++i) {
                const instanceIndex = Math.floor(i / cellCount);
                const groupIndex = i % cellCount;
                const vs = getIndices(instanceIndex, groupIndex);
                if (vs !== undefined) {
                    for (const v of vs) {
                        if (apply(interval_1.Interval.ofSingleton(v)))
                            changed = true;
                    }
                }
            }
            return changed;
        });
    }
    else {
        return (0, util_2.eachVolumeLoci)(loci, volume, undefined, apply);
    }
}
//
function SliceVisual(materialId, volume, key) {
    return (0, visual_1.VolumeVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.SliceParams),
        createGeometry: createImage,
        createLocationIterator: util_2.createVolumeSliceCellLocationIterator,
        getLoci: getSliceLoci,
        eachLocation: eachSlice,
        setUpdateState: (state, newVolume, currentVolume, newProps, currentProps, newTheme, currentTheme) => {
            state.createGeometry = (newProps.dimension.name !== currentProps.dimension.name ||
                newProps.dimension.params !== currentProps.dimension.params ||
                newProps.mode !== currentProps.mode ||
                !vec3_1.Vec3.equals(newProps.rotation.axis, currentProps.rotation.axis) ||
                newProps.rotation.angle !== currentProps.rotation.angle ||
                newProps.offset !== currentProps.offset ||
                newProps.axis !== currentProps.axis ||
                !vec3_1.Vec3.equals(newProps.plane.point, currentProps.plane.point) ||
                !vec3_1.Vec3.equals(newProps.plane.normal, currentProps.plane.normal) ||
                !volume_1.Volume.IsoValue.areSame(newProps.isoValue, currentProps.isoValue, newVolume.grid.stats) ||
                !color_2.ColorTheme.areEqual(newTheme.color, currentTheme.color) ||
                !volume_1.Volume.areInstanceTransformsEqual(newVolume, currentVolume));
        },
        geometryUtils: image_1.Image.Utils
    }, materialId);
}
function SliceRepresentation(ctx, getParams) {
    return (0, representation_1.VolumeRepresentation)('Slice', ctx, getParams, SliceVisual, getLoci);
}
exports.SliceRepresentationProvider = (0, representation_1.VolumeRepresentationProvider)({
    name: 'slice',
    label: 'Slice',
    description: 'Slice of volume rendered as image with interpolation.',
    factory: SliceRepresentation,
    getParams: getSliceParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.SliceParams),
    defaultColorTheme: { name: 'uniform' },
    defaultSizeTheme: { name: 'uniform' },
    isApplicable: (volume) => !volume_1.Volume.isEmpty(volume) && !volume_1.Volume.Segmentation.get(volume),
});
