// molstar/lib/commonjs/mol-geo/geometry/interior.js
"use strict";
/**
 * Copyright (c) 2025-2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInteriorParam = getInteriorParam;
exports.areInteriorPropsEquals = areInteriorPropsEquals;
exports.getInteriorColor = getInteriorColor;
exports.getInteriorSubstance = getInteriorSubstance;
exports.createInteriorValues = createInteriorValues;
exports.updateInteriorValues = updateInteriorValues;
const vec4_1 = require("../../mol-math/linear-algebra/3d/vec4.js");
const color_1 = require("../../mol-util/color/color.js");
const material_1 = require("../../mol-util/material.js");
const param_definition_1 = require("../../mol-util/param-definition.js");
const value_cell_1 = require("../../mol-util/value-cell.js");
function getInteriorParam() {
    return param_definition_1.ParamDefinition.Group({
        color: param_definition_1.ParamDefinition.Color(color_1.Color.fromRgb(76, 76, 76)),
        colorStrength: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 1, step: 0.01 }),
        substance: material_1.Material.getParam(),
        substanceStrength: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 1, step: 0.01 }),
    });
}
function areInteriorPropsEquals(a, b) {
    return a.color === b.color
        && a.colorStrength === b.colorStrength
        && material_1.Material.areEqual(a.substance, b.substance)
        && a.substanceStrength === b.substanceStrength;
}
function getInteriorColor(props, out) {
    color_1.Color.toArrayNormalized(props.color, out, 0);
    out[3] = props.colorStrength;
    return out;
}
function getInteriorSubstance(props, out) {
    material_1.Material.toArrayNormalized(props.substance, out, 0);
    out[3] = props.substanceStrength;
    return out;
}
function createInteriorValues(props) {
    return {
        uInteriorColor: value_cell_1.ValueCell.create(getInteriorColor(props, (0, vec4_1.Vec4)())),
        uInteriorSubstance: value_cell_1.ValueCell.create(getInteriorSubstance(props, (0, vec4_1.Vec4)())),
    };
}
function updateInteriorValues(values, props) {
    value_cell_1.ValueCell.update(values.uInteriorColor, getInteriorColor(props, values.uInteriorColor.ref.value));
    value_cell_1.ValueCell.update(values.uInteriorSubstance, getInteriorSubstance(props, values.uInteriorSubstance.ref.value));
}
