// molstar/lib/commonjs/extensions/mvs/load-extensions/volume-streaming.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VolumeStreamingExtension = void 0;
const transformers_1 = require("../../../mol-plugin/behavior/dynamic/volume-streaming/transformers.js");
const object_1 = require("../../../mol-util/object.js");
const utils_1 = require("../helpers/utils.js");
const load_generic_1 = require("../load-generic.js");
/** This MVS loading extension allows turning on volume streaming for a structure by providing custom property `molstar_volume_streaming`.
 *
 *  Examples:
 *
 *  ```
 *  builder
 *      .download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1cbs_updated.cif' })
 *      .parse({ format: 'mmcif' })
 *      .modelStructure({
 *          custom: {
 *              molstar_volume_streaming: true,
 *          },
 *      })
 *      .component()
 *      .representation();
 *
 *  builder
 *      .download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1tqn_updated.cif' })
 *      .parse({ format: 'mmcif' })
 *      .modelStructure({
 *          custom: {
 *              molstar_volume_streaming: {
 *                  channel_params: {
 *                      '2fo-fc': { color: 'skyblue', opacity: 0.3 },
 *                      'fo-fc(+ve)': { color: 'greenyellow', wireframe: true, isoValue: { kind: 'relative', relativeValue: +2.5 } },
 *                      'fo-fc(-ve)': { color: 'orange', wireframe: true, isoValue: { kind: 'relative', relativeValue: -2.5 } },
 *                  },
 *              } satisfies MolstarVolumeStreamingCustomProp,
 *          },
 *      })
 *      .component()
 *      .representation();
 *
 *  builder
 *      .download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/8hra_updated.cif' })
 *      .parse({ format: 'mmcif' })
 *      .modelStructure({
 *          custom: {
 *              molstar_volume_streaming: {
 *                  server_url: 'https://www.ebi.ac.uk/pdbe/densities', // = default
 *                  entries: [{ dataId: 'EMD-34965', source: { name: 'em', params: { isoValue: { kind: 'absolute', absoluteValue: 0.015 } } } }],
 *                  view: 'auto', // default is 'auto' for EM, 'selection-box' for X-ray structures
 *                  channel_params: {
 *                      em: { color: '#ff0000', opacity: 0.4, isoValue: { kind: 'absolute', absoluteValue: 0.025 } },
 *                  },
 *              } satisfies MolstarVolumeStreamingCustomProp,
 *          },
 *      })
 *      .component()
 *      .representation();
 *  ```
 */
exports.VolumeStreamingExtension = {
    id: 'wwpdb/volume-streaming',
    description: 'Allow turning on volume streaming for a structure',
    createExtensionContext: () => ({}),
    action: (updateTarget, node, context, extContext) => {
        var _a;
        if (node.kind !== 'structure')
            return;
        let params = (_a = node.custom) === null || _a === void 0 ? void 0 : _a.molstar_volume_streaming;
        if (!params)
            return;
        if (params === true)
            params = {};
        const streamingInfo = load_generic_1.UpdateTarget.apply(updateTarget, transformers_1.CreateVolumeStreamingInfo, {
            serverUrl: params.server_url,
            autoEntries: !params.entries,
            entries: params.entries,
            defaultView: params.view,
            defaultChannelParams: params.channel_params && (0, object_1.mapObjectMap)(params.channel_params, normalizeChannelParams),
        }, { state: { isCollapsed: true } });
        const streamingBehavior = load_generic_1.UpdateTarget.apply(streamingInfo, transformers_1.CreateVolumeStreamingBehavior);
        load_generic_1.UpdateTarget.apply(streamingBehavior, transformers_1.VolumeStreamingVisual, { channel: '2fo-fc' }, { state: { isGhost: true }, tags: '2fo-fc' });
        load_generic_1.UpdateTarget.apply(streamingBehavior, transformers_1.VolumeStreamingVisual, { channel: 'fo-fc(+ve)' }, { state: { isGhost: true }, tags: 'fo-fc(+ve)' });
        load_generic_1.UpdateTarget.apply(streamingBehavior, transformers_1.VolumeStreamingVisual, { channel: 'fo-fc(-ve)' }, { state: { isGhost: true }, tags: 'fo-fc(-ve)' });
        load_generic_1.UpdateTarget.apply(streamingBehavior, transformers_1.VolumeStreamingVisual, { channel: 'em' }, { state: { isGhost: true }, tags: 'em' });
    },
};
function normalizeChannelParams(p) {
    if (!p)
        return undefined;
    return {
        ...p,
        color: (0, utils_1.decodeColor)(p.color),
    };
}
