// molstar/lib/apps/viewer/extensions.js
/**
 * Copyright (c) 2018-2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author Adam Midlik <midlik@gmail.com>
 */
import { ANVILMembraneOrientation } from '../../extensions/anvil/behavior.js';
import { AssemblySymmetry } from '../../extensions/assembly-symmetry/index.js';
import { Backgrounds } from '../../extensions/backgrounds/index.js';
import { DebugHelpers } from '../../extensions/debug-helpers/index.js';
import { DnatcoNtCs } from '../../extensions/dnatco/index.js';
import { G3DFormat } from '../../extensions/g3d/format.js';
import { GeometryExport } from '../../extensions/geo-export/index.js';
import { MAQualityAssessment, MAQualityAssessmentConfig } from '../../extensions/model-archive/quality-assessment/behavior.js';
import { ModelExport } from '../../extensions/model-export/index.js';
import { Mp4Export } from '../../extensions/mp4-export/index.js';
import { loadMVS } from '../../extensions/mvs/index.js';
import { MolViewSpec } from '../../extensions/mvs/behavior.js';
import { loadMVSData } from '../../extensions/mvs/components/formats.js';
import { PDBeStructureQualityReport } from '../../extensions/pdbe/index.js';
import { RCSBValidationReport } from '../../extensions/rcsb/index.js';
import { SbNcbrPartialCharges, SbNcbrTunnels } from '../../extensions/sb-ncbr/index.js';
import { wwPDBChemicalComponentDictionary } from '../../extensions/wwpdb/ccd/behavior.js';
import { wwPDBStructConnExtensionFunctions } from '../../extensions/wwpdb/struct-conn/index.js';
import { ZenodoImport } from '../../extensions/zenodo/index.js';
import { PluginSpec } from '../../mol-plugin/spec.js';
import { MVSData } from '../../extensions/mvs/mvs-data.js';
import * as MVSUtil from '../../extensions/mvs/util.js';
export const ExtensionMap = {
    // Mol* built-in extensions
    'mvs': PluginSpec.Behavior(MolViewSpec),
    'backgrounds': PluginSpec.Behavior(Backgrounds),
    'debug-helpers': PluginSpec.Behavior(DebugHelpers),
    'model-export': PluginSpec.Behavior(ModelExport),
    'mp4-export': PluginSpec.Behavior(Mp4Export),
    'geo-export': PluginSpec.Behavior(GeometryExport),
    'zenodo-import': PluginSpec.Behavior(ZenodoImport),
    'wwpdb-chemical-component-dictionary': PluginSpec.Behavior(wwPDBChemicalComponentDictionary),
    // 3rd party extensions
    'pdbe-structure-quality-report': PluginSpec.Behavior(PDBeStructureQualityReport),
    'dnatco-ntcs': PluginSpec.Behavior(DnatcoNtCs),
    'assembly-symmetry': PluginSpec.Behavior(AssemblySymmetry),
    'rcsb-validation-report': PluginSpec.Behavior(RCSBValidationReport),
    'anvil-membrane-orientation': PluginSpec.Behavior(ANVILMembraneOrientation),
    'g3d': PluginSpec.Behavior(G3DFormat), // TODO: consider removing this for Mol* 6.0
    'ma-quality-assessment': PluginSpec.Behavior(MAQualityAssessment),
    'sb-ncbr-partial-charges': PluginSpec.Behavior(SbNcbrPartialCharges),
    'tunnels': PluginSpec.Behavior(SbNcbrTunnels),
};
export const PluginExtensions = {
    wwPDBStructConn: wwPDBStructConnExtensionFunctions,
    mvs: {
        MVSData,
        createBuilder: MVSData.createBuilder,
        loadMVS,
        loadMVSData,
        util: {
            ...MVSUtil
        }
    },
    modelArchive: {
        qualityAssessment: {
            config: MAQualityAssessmentConfig
        }
    }
};
