// molstar/lib/commonjs/extensions/debug-helpers/clip-object-helper.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClipObjectHelper = exports.ClipObjectHelperParams = void 0;
const render_object_1 = require("../../mol-gl/render-object.js");
const mesh_builder_1 = require("../../mol-geo/geometry/mesh/mesh-builder.js");
const sphere_1 = require("../../mol-geo/geometry/mesh/builder/sphere.js");
const cylinder_1 = require("../../mol-geo/geometry/mesh/builder/cylinder.js");
const mesh_1 = require("../../mol-geo/geometry/mesh/mesh.js");
const param_definition_1 = require("../../mol-util/param-definition.js");
const scene_1 = require("../../mol-gl/scene.js");
const names_1 = require("../../mol-util/color/names.js");
const clip_1 = require("../../mol-util/clip.js");
const mat4_1 = require("../../mol-math/linear-algebra/3d/mat4.js");
const quat_1 = require("../../mol-math/linear-algebra/3d/quat.js");
const vec3_1 = require("../../mol-math/linear-algebra/3d/vec3.js");
const box_1 = require("../../mol-geo/primitive/box.js");
const plane_1 = require("../../mol-geo/primitive/plane.js");
const cylinder_2 = require("../../mol-geo/primitive/cylinder.js");
const sphere_2 = require("../../mol-geo/primitive/sphere.js");
exports.ClipObjectHelperParams = {
    clipObjects: param_definition_1.ParamDefinition.Boolean(false, { description: 'Show clip-objects of visible render objects.' }),
};
//
/** Serializes clip object params to a string key for deduplication */
function clipObjectKey(type, invert, position, posOffset, rotation, rotOffset, scale, scaleOffset, transform, transformOffset) {
    // Round floats to 5 decimal places to avoid floating point noise
    const r = (v) => Math.round(v * 100000) / 100000;
    const parts = [
        type, invert ? 1 : 0,
        r(position[posOffset]), r(position[posOffset + 1]), r(position[posOffset + 2]),
        r(rotation[rotOffset]), r(rotation[rotOffset + 1]), r(rotation[rotOffset + 2]), r(rotation[rotOffset + 3]),
        r(scale[scaleOffset]), r(scale[scaleOffset + 1]), r(scale[scaleOffset + 2]),
    ];
    for (let j = 0; j < 16; ++j) {
        parts.push(r(transform[transformOffset + j]));
    }
    return parts.join(',');
}
const clipObjectColors = {
    [clip_1.Clip.Type.plane]: names_1.ColorNames.orange,
    [clip_1.Clip.Type.sphere]: names_1.ColorNames.green,
    [clip_1.Clip.Type.cube]: names_1.ColorNames.dodgerblue,
    [clip_1.Clip.Type.cylinder]: names_1.ColorNames.gold,
    [clip_1.Clip.Type.infiniteCone]: names_1.ColorNames.crimson,
};
const clipMaterialId = (0, render_object_1.getNextMaterialId)();
const indicatorMaterialId = (0, render_object_1.getNextMaterialId)();
// Pre-rotation matrices for aligning primitives to GLSL SDF local frames
// Plane: Rx(-90°) maps primitive Z-normal to GLSL Y-normal
const preRotPlaneQuat = quat_1.Quat.setAxisAngle((0, quat_1.Quat)(), vec3_1.Vec3.create(1, 0, 0), -Math.PI / 2);
const preRotPlaneMat = mat4_1.Mat4.fromQuat((0, mat4_1.Mat4)(), preRotPlaneQuat);
// Cone: Rx(+90°) maps primitive Y-axis to GLSL Z-axis
const preRotConeQuat = quat_1.Quat.setAxisAngle((0, quat_1.Quat)(), vec3_1.Vec3.create(1, 0, 0), Math.PI / 2);
const preRotConeMat = mat4_1.Mat4.fromQuat((0, mat4_1.Mat4)(), preRotConeQuat);
// Temp variables for constructing transforms
const _position = (0, vec3_1.Vec3)();
const _rotation = (0, quat_1.Quat)();
const _scale = (0, vec3_1.Vec3)();
const _clipTransform = (0, mat4_1.Mat4)();
const _invClipTransform = (0, mat4_1.Mat4)();
const _rotMat = (0, mat4_1.Mat4)();
const _translateMat = (0, mat4_1.Mat4)();
const _baseMat = (0, mat4_1.Mat4)();
const _tmpMat = (0, mat4_1.Mat4)();
const _axisEnd = (0, vec3_1.Vec3)();
const _yAxis = vec3_1.Vec3.create(0, 1, 0);
const _zAxis = vec3_1.Vec3.create(0, 0, 1);
const _indicatorPos = (0, vec3_1.Vec3)();
class ClipObjectHelper {
    constructor(ctx, parent, props) {
        this.objectsData = new Map();
        this.scene = scene_1.Scene.create(ctx, 'blended');
        this.parent = parent;
        this._props = { ...param_definition_1.ParamDefinition.getDefaultValues(exports.ClipObjectHelperParams), ...props };
    }
    update() {
        const currentKeys = new Set();
        const sceneRadius = this.parent.boundingSphereVisible.radius || 50;
        this.parent.forEach((r, ro) => {
            if (!ro.state.visible)
                return;
            const count = ro.values.dClipObjectCount.ref.value;
            if (count === 0)
                return;
            const types = ro.values.uClipObjectType.ref.value;
            const inverts = ro.values.uClipObjectInvert.ref.value;
            const positions = ro.values.uClipObjectPosition.ref.value;
            const rotations = ro.values.uClipObjectRotation.ref.value;
            const scales = ro.values.uClipObjectScale.ref.value;
            const transforms = ro.values.uClipObjectTransform.ref.value;
            for (let i = 0; i < count; ++i) {
                const type = types[i];
                if (type === clip_1.Clip.Type.none)
                    continue;
                const key = clipObjectKey(type, inverts[i], positions, i * 3, rotations, i * 4, scales, i * 3, transforms, i * 16);
                currentKeys.add(key);
                if (this.objectsData.has(key))
                    continue;
                // Extract per-object params
                vec3_1.Vec3.fromArray(_position, positions, i * 3);
                quat_1.Quat.fromArray(_rotation, rotations, i * 4);
                quat_1.Quat.normalize(_rotation, _rotation); // ensure unit quaternion for proper rotation
                vec3_1.Vec3.fromArray(_scale, scales, i * 3);
                mat4_1.Mat4.fromArray(_clipTransform, transforms, i * 16);
                // Build base transform (translate * rotate) without scale,
                // so each shape can insert pre-rotations before scale.
                mat4_1.Mat4.fromQuat(_rotMat, _rotation);
                mat4_1.Mat4.fromTranslation(_translateMat, _position);
                mat4_1.Mat4.mul(_baseMat, _translateMat, _rotMat);
                // apply inverse of clip transform
                if (!mat4_1.Mat4.isIdentity(_clipTransform)) {
                    mat4_1.Mat4.invert(_invClipTransform, _clipTransform);
                    mat4_1.Mat4.mul(_baseMat, _invClipTransform, _baseMat);
                }
                const mesh = createClipObjectMesh(type, _baseMat, _scale, sceneRadius);
                const color = clipObjectColors[type] || names_1.ColorNames.white;
                const renderObject = createClipObjectRenderObject(mesh, color, clipMaterialId, type);
                // Create position/rotation indicator mesh
                const invert = inverts[i];
                const indicatorMesh = createIndicatorMesh(_position, _rotation, _clipTransform, _scale, type, invert);
                const indicatorRenderObject = createIndicatorRenderObject(indicatorMesh, indicatorMaterialId);
                this.scene.add(renderObject);
                this.scene.add(indicatorRenderObject);
                this.objectsData.set(key, { key, renderObject, indicatorRenderObject, mesh });
            }
        });
        // Remove clip objects no longer present
        this.objectsData.forEach((data, key) => {
            if (!currentKeys.has(key)) {
                this.scene.remove(data.renderObject);
                this.scene.remove(data.indicatorRenderObject);
                this.objectsData.delete(key);
            }
        });
        this.scene.update(void 0, false);
        this.scene.commit();
    }
    syncVisibility() {
        const visible = this._props.clipObjects;
        this.objectsData.forEach(data => {
            data.renderObject.state.visible = visible;
            data.indicatorRenderObject.state.visible = visible;
        });
    }
    clear() {
        this.objectsData.clear();
        this.scene.clear();
    }
    get isEnabled() {
        return this._props.clipObjects;
    }
    get props() { return this._props; }
    setProps(props) {
        Object.assign(this._props, props);
        if (this.isEnabled)
            this.update();
    }
}
exports.ClipObjectHelper = ClipObjectHelper;
//
function createClipObjectMesh(type, baseMat, scale, sceneRadius) {
    switch (type) {
        case clip_1.Clip.Type.plane: return createPlaneMesh(baseMat, sceneRadius);
        case clip_1.Clip.Type.sphere: return createSphereMesh(baseMat, scale);
        case clip_1.Clip.Type.cube: return createCubeMesh(baseMat, scale);
        case clip_1.Clip.Type.cylinder: return createCylinderMesh(baseMat, scale);
        case clip_1.Clip.Type.infiniteCone: return createConeMesh(baseMat, scale, sceneRadius);
        default: return createSphereMesh(baseMat, scale); // fallback
    }
}
/**
 * Plane: GLSL normal is quaternionTransform(rotation, vec3(0,1,0)) — Y-up default.
 * Plane() primitive lies in XY with normal (0,0,1) along Z.
 * Pre-rotate Rx(-90°) to align primitive Z-normal to GLSL Y-normal.
 * Sized to cover the scene bounding sphere. Clip scale is ignored (plane is infinite in GLSL).
 */
function createPlaneMesh(baseMat, sceneRadius) {
    const size = Math.max(sceneRadius * 2, 10);
    // baseMat * preRotPlane * uniformScale(size)
    mat4_1.Mat4.mul(_tmpMat, baseMat, preRotPlaneMat);
    mat4_1.Mat4.scale(_tmpMat, _tmpMat, vec3_1.Vec3.create(size, size, 1));
    const plane = (0, plane_1.Plane)();
    const builderState = mesh_builder_1.MeshBuilder.createState(256, 128);
    mesh_builder_1.MeshBuilder.addPrimitive(builderState, _tmpMat, plane);
    // Add flipped backface for double-sided visibility
    mesh_builder_1.MeshBuilder.addPrimitiveFlipped(builderState, _tmpMat, plane);
    return mesh_builder_1.MeshBuilder.getMesh(builderState);
}
/**
 * Sphere: SDF uses scale * 0.5 as the radii (ellipsoid).
 * Sphere primitive has radius 1.
 * Transform: baseMat * scale * 0.5
 */
function createSphereMesh(baseMat, scale) {
    const detail = 2;
    const sphere = getSphereForHelper(detail);
    // baseMat * scale(scale * 0.5)
    mat4_1.Mat4.scale(_tmpMat, baseMat, vec3_1.Vec3.create(scale[0] * 0.5, scale[1] * 0.5, scale[2] * 0.5));
    const vertexCount = 10 * Math.pow(2, 2 * detail) + 2;
    const builderState = mesh_builder_1.MeshBuilder.createState(vertexCount * 3, vertexCount);
    mesh_builder_1.MeshBuilder.addPrimitive(builderState, _tmpMat, sphere);
    return mesh_builder_1.MeshBuilder.getMesh(builderState);
}
let _helperSphere;
function getSphereForHelper(detail) {
    if (!_helperSphere)
        _helperSphere = (0, sphere_2.Sphere)(detail);
    return _helperSphere;
}
/**
 * Cube: SDF uses scale * 0.5 as half-extents.
 * Box() primitive is ±0.5 (unit cube), so scaling by `scale` gives half-extents of scale*0.5.
 */
function createCubeMesh(baseMat, scale) {
    // baseMat * scale(scale)
    mat4_1.Mat4.scale(_tmpMat, baseMat, scale);
    const box = (0, box_1.Box)();
    const builderState = mesh_builder_1.MeshBuilder.createState(256, 128);
    mesh_builder_1.MeshBuilder.addPrimitive(builderState, _tmpMat, box);
    return mesh_builder_1.MeshBuilder.getMesh(builderState);
}
/**
 * Cylinder: SDF axis along Y, radius = scale.x * 0.5, half-height = scale.y * 0.5.
 * Cylinder primitive: axis along Y, radius=1 in XZ, half-height=0.5 in Y.
 * Need: X/Z *= scale.x * 0.5 (radius 1 → scale.x*0.5), Y *= scale.y (half-height 0.5 → scale.y*0.5).
 */
function createCylinderMesh(baseMat, scale) {
    const cyl = (0, cylinder_2.Cylinder)({ radiusTop: 1, radiusBottom: 1, height: 1, radialSegments: 16, heightSegments: 1, topCap: true, bottomCap: true });
    // baseMat * scale(scale.x * 0.5, scale.y, scale.x * 0.5) — use scale.x for both radial axes
    mat4_1.Mat4.scale(_tmpMat, baseMat, vec3_1.Vec3.create(scale[0] * 0.5, scale[1], scale[0] * 0.5));
    const vertexCount = cyl.vertices.length / 3;
    const builderState = mesh_builder_1.MeshBuilder.createState(vertexCount * 3, vertexCount);
    mesh_builder_1.MeshBuilder.addPrimitive(builderState, _tmpMat, cyl);
    return mesh_builder_1.MeshBuilder.getMesh(builderState);
}
/**
 * InfiniteCone: GLSL SDF axis along Z, radial in XY.
 *   surface: size.x * length(t.xy) + size.y * t.z = 0  (size = scale * 0.5)
 *   half-angle: tan(θ) = scale.y / scale.x
 *   apex at clip position (origin), opens in -Z direction.
 *
 * Cylinder primitive (radiusTop=0, radiusBottom=1, height=1):
 *   axis along Y, tip at Y=+0.5, base at Y=-0.5, base radius=1.
 *
 * Transform chain (right-to-left):
 *   1. Scale(baseRadius, coneLength, baseRadius): stretch primitive to correct proportions
 *   2. Translate(0, -0.5*coneLength, 0): move tip from Y=+0.5·cL to Y=0 (apex at origin)
 *      (after scale, tip is at Y=+0.5·cL; shifting by -0.5·cL puts it at Y=0)
 *   3. preRotCone Rx(+90°): map prim-Y→Z, so cone axis becomes Z, opening in -Z
 *   4. baseMat: position + rotation of clip object
 */
function createConeMesh(baseMat, scale, sceneRadius) {
    const cone = (0, cylinder_2.Cylinder)({ radiusTop: 0, radiusBottom: 1, height: 1, radialSegments: 16, heightSegments: 1, topCap: false, bottomCap: true });
    // Visible length of the (infinite) cone, and base radius matching the GLSL half-angle
    const coneLength = Math.max(sceneRadius * 2, 10);
    const tanHalfAngle = (scale[1] || 1) / (scale[0] || 1); // tan(θ) = scaleY / scaleX
    const baseRadius = coneLength * tanHalfAngle;
    // baseMat * preRotCone * Translate(0, -coneLength/2, 0) * Scale(baseRadius, coneLength, baseRadius)
    const scaleMat = mat4_1.Mat4.fromScaling((0, mat4_1.Mat4)(), vec3_1.Vec3.create(baseRadius, coneLength, baseRadius));
    const translateMat = mat4_1.Mat4.fromTranslation((0, mat4_1.Mat4)(), vec3_1.Vec3.create(0, -coneLength * 0.5, 0));
    mat4_1.Mat4.mul(_tmpMat, translateMat, scaleMat);
    mat4_1.Mat4.mul(_tmpMat, preRotConeMat, _tmpMat);
    mat4_1.Mat4.mul(_tmpMat, baseMat, _tmpMat);
    const vertexCount = cone.vertices.length / 3;
    const builderState = mesh_builder_1.MeshBuilder.createState(vertexCount * 3, vertexCount);
    mesh_builder_1.MeshBuilder.addPrimitive(builderState, _tmpMat, cone);
    return mesh_builder_1.MeshBuilder.getMesh(builderState);
}
function createClipObjectRenderObject(mesh, color, materialId, type) {
    const alpha = type === clip_1.Clip.Type.plane ? 0.25 : 0.15;
    const values = mesh_1.Mesh.Utils.createValuesSimple(mesh, { alpha, doubleSided: false, cellSize: 0, batchSize: 0 }, color, 1);
    return (0, render_object_1.createRenderObject)('mesh', values, { disposed: false, visible: true, alphaFactor: 1, pickable: false, colorOnly: false, opaque: false, writeDepth: false }, materialId);
}
/**
 * Create a mesh with a sphere at the clip object position and a cylinder
 * along the characteristic axis to indicate orientation.
 *
 * - Plane/sphere/cube/cylinder: axis = rotated Y (matches GLSL normal/axis direction)
 * - InfiniteCone: axis = rotated Z (cone axis is Z in local frame)
 * - Plane with invert: direction is flipped
 */
function createIndicatorMesh(position, rotation, clipTransform, scale, type, invert) {
    const objectSize = Math.max(scale[0], scale[1], scale[2]);
    const sphereRadius = Math.max(objectSize * 0.004, 0.01);
    const cylinderRadius = sphereRadius * 0.4;
    const axisLength = Math.max(objectSize * 0.1, 2);
    // Transform position by inverse of clipTransform if non-identity
    vec3_1.Vec3.copy(_indicatorPos, position);
    if (!mat4_1.Mat4.isIdentity(clipTransform)) {
        mat4_1.Mat4.invert(_invClipTransform, clipTransform);
        vec3_1.Vec3.transformMat4(_indicatorPos, _indicatorPos, _invClipTransform);
    }
    // Choose the local-frame axis based on clip type
    const localAxis = type === clip_1.Clip.Type.infiniteCone ? _zAxis : _yAxis;
    vec3_1.Vec3.transformQuat(_axisEnd, localAxis, rotation);
    // Cone opens in -Z locally, so negate to point along the cone opening
    if (type === clip_1.Clip.Type.infiniteCone) {
        vec3_1.Vec3.negate(_axisEnd, _axisEnd);
    }
    // For planes, the normal points toward the clipped (removed) side.
    // Flip so the indicator points toward the non-clipped (kept) geometry.
    // When inverted, the kept side is the normal side, so don't flip.
    if (type === clip_1.Clip.Type.plane && !invert) {
        vec3_1.Vec3.negate(_axisEnd, _axisEnd);
    }
    // If clipTransform is non-identity, also transform the axis direction
    if (!mat4_1.Mat4.isIdentity(clipTransform)) {
        // Transform direction (not position) by inverse clipTransform
        const endWorld = (0, vec3_1.Vec3)();
        vec3_1.Vec3.add(endWorld, position, vec3_1.Vec3.scale((0, vec3_1.Vec3)(), _axisEnd, axisLength));
        vec3_1.Vec3.transformMat4(endWorld, endWorld, _invClipTransform);
        vec3_1.Vec3.sub(_axisEnd, endWorld, _indicatorPos);
        vec3_1.Vec3.normalize(_axisEnd, _axisEnd);
    }
    // Axis cylinder endpoint
    const axisEndPoint = (0, vec3_1.Vec3)();
    vec3_1.Vec3.scaleAndAdd(axisEndPoint, _indicatorPos, _axisEnd, axisLength);
    const builderState = mesh_builder_1.MeshBuilder.createState(512, 256);
    // Position sphere
    (0, sphere_1.addSphere)(builderState, _indicatorPos, sphereRadius, 1);
    // Rotation axis cylinder
    (0, cylinder_1.addCylinder)(builderState, _indicatorPos, axisEndPoint, 1, { radiusTop: cylinderRadius, radiusBottom: cylinderRadius, radialSegments: 8 });
    // Small sphere at tip of axis
    (0, sphere_1.addSphere)(builderState, axisEndPoint, cylinderRadius * 1.5, 1);
    return mesh_builder_1.MeshBuilder.getMesh(builderState);
}
function createIndicatorRenderObject(mesh, materialId) {
    const values = mesh_1.Mesh.Utils.createValuesSimple(mesh, { alpha: 0.7, doubleSided: false, cellSize: 0, batchSize: 0 }, names_1.ColorNames.white, 1);
    return (0, render_object_1.createRenderObject)('mesh', values, { disposed: false, visible: true, alphaFactor: 1, pickable: false, colorOnly: false, opaque: false, writeDepth: false }, materialId);
}
