// molstar/lib/commonjs/extensions/debug-helpers/direct-volume-helper.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectVolumeHelper = exports.DirectVolumeHelperParams = void 0;
const render_object_1 = require("../../mol-gl/render-object.js");
const param_definition_1 = require("../../mol-util/param-definition.js");
const scene_1 = require("../../mol-gl/scene.js");
const names_1 = require("../../mol-util/color/names.js");
const lines_1 = require("../../mol-geo/geometry/lines/lines.js");
const lines_builder_1 = require("../../mol-geo/geometry/lines/lines-builder.js");
const mat4_1 = require("../../mol-math/linear-algebra/3d/mat4.js");
const box_1 = require("../../mol-geo/geometry/lines/builder/box.js");
exports.DirectVolumeHelperParams = {
    directVolumeEdges: param_definition_1.ParamDefinition.Boolean(false, { description: 'Show edges of visible direct-volume render objects.' }),
};
const directVolumeMaterialId = (0, render_object_1.getNextMaterialId)();
class DirectVolumeHelper {
    constructor(ctx, parent, props) {
        this.renderObjects = new Map();
        this.scene = scene_1.Scene.create(ctx, 'blended');
        this.parent = parent;
        this._props = { ...param_definition_1.ParamDefinition.getDefaultValues(exports.DirectVolumeHelperParams), ...props };
    }
    update() {
        const previousIds = new Set(this.renderObjects.keys());
        this.parent.forEach((r, ro) => {
            if (!ro.state.visible)
                return;
            if (ro.type !== 'direct-volume')
                return;
            const values = ro.values;
            const version = values.uUnitToCartn.ref.version + values.uGridDim.ref.version + values.aTransform.ref.version;
            const existing = this.renderObjects.get(ro.id);
            if (existing && existing.version === version) {
                previousIds.delete(ro.id);
                return;
            }
            // Remove old entry if version changed
            if (existing) {
                this.scene.remove(existing.ro);
                this.renderObjects.delete(ro.id);
            }
            const lines = createVolumeEdgeLines(values);
            if (!lines)
                return;
            const linesRO = createLinesRenderObject(lines, directVolumeMaterialId);
            this.scene.add(linesRO);
            this.renderObjects.set(ro.id, { ro: linesRO, version });
            previousIds.delete(ro.id);
        });
        for (const id of previousIds) {
            const entry = this.renderObjects.get(id);
            if (entry) {
                this.scene.remove(entry.ro);
                this.renderObjects.delete(id);
            }
        }
        this.scene.update(void 0, false);
        this.scene.commit();
    }
    syncVisibility() {
        const visible = this._props.directVolumeEdges;
        this.renderObjects.forEach(entry => {
            entry.ro.state.visible = visible;
        });
    }
    clear() {
        this.renderObjects.clear();
        this.scene.clear();
    }
    get isEnabled() {
        return this._props.directVolumeEdges;
    }
    get props() { return this._props; }
    setProps(props) {
        Object.assign(this._props, props);
        if (this.isEnabled)
            this.update();
    }
}
exports.DirectVolumeHelper = DirectVolumeHelper;
//
/**
 * The volume proxy box in the shader uses aPosition in [-0.5, 0.5]^3,
 * shifted to [0,1]^3 (unitCoord = aPosition + 0.5), then transformed by:
 *   uUnitToCartn → Cartesian space
 *   aTransform → instance space
 *
 * We replicate this pipeline to get the correct world-space edges.
 * Grid ticks are placed at 1/gridDim intervals along each edge.
 */
function createVolumeEdgeLines(values) {
    const unitToCartn = values.uUnitToCartn.ref.value;
    const transforms = values.aTransform.ref.value;
    const instanceCount = values.uInstanceCount.ref.value;
    const bs = values.boundingSphere.ref.value;
    if (bs.radius < 1e-6)
        return undefined;
    const builder = lines_builder_1.LinesBuilder.create(128 * instanceCount);
    for (let inst = 0; inst < instanceCount; ++inst) {
        const instTransform = (0, mat4_1.Mat4)();
        mat4_1.Mat4.fromArray(instTransform, transforms, inst * 16);
        // Combined transform: aTransform * uUnitToCartn
        const combined = mat4_1.Mat4.mul((0, mat4_1.Mat4)(), instTransform, unitToCartn);
        (0, box_1.addBox)(builder, combined, 0);
    }
    return builder.getLines();
}
function createLinesRenderObject(lines, materialId) {
    const props = { ...param_definition_1.ParamDefinition.getDefaultValues(lines_1.Lines.Params), sizeFactor: 1, alpha: 0.8 };
    const values = lines_1.Lines.Utils.createValuesSimple(lines, props, names_1.ColorNames.orange, 1);
    const state = lines_1.Lines.Utils.createRenderableState(props);
    state.pickable = false;
    return (0, render_object_1.createRenderObject)('lines', values, state, materialId);
}
