// molstar/lib/commonjs/extensions/debug-helpers/image-helper.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageHelper = exports.ImageHelperParams = void 0;
const render_object_1 = require("../../mol-gl/render-object.js");
const param_definition_1 = require("../../mol-util/param-definition.js");
const scene_1 = require("../../mol-gl/scene.js");
const names_1 = require("../../mol-util/color/names.js");
const lines_1 = require("../../mol-geo/geometry/lines/lines.js");
const lines_builder_1 = require("../../mol-geo/geometry/lines/lines-builder.js");
const vec3_1 = require("../../mol-math/linear-algebra/3d/vec3.js");
const mat4_1 = require("../../mol-math/linear-algebra/3d/mat4.js");
const quat_1 = require("../../mol-math/linear-algebra/3d/quat.js");
const clip_1 = require("../../mol-util/clip.js");
const sphere_1 = require("../../mol-geo/geometry/lines/builder/sphere.js");
const box_1 = require("../../mol-geo/geometry/lines/builder/box.js");
const plane_1 = require("../../mol-geo/geometry/lines/builder/plane.js");
exports.ImageHelperParams = {
    imageEdges: param_definition_1.ParamDefinition.Boolean(false, { description: 'Show edges of visible image render objects.' }),
};
const imageEdgeMaterialId = (0, render_object_1.getNextMaterialId)();
const imageTrimMaterialId = (0, render_object_1.getNextMaterialId)();
// Temp vectors
const _trimPos = (0, vec3_1.Vec3)();
const _trimScale = (0, vec3_1.Vec3)();
const _trimRot = (0, quat_1.Quat)();
const _trimTransform = (0, mat4_1.Mat4)();
const _tmpMat = (0, mat4_1.Mat4)();
class ImageHelper {
    constructor(ctx, parent, props) {
        this.renderObjects = new Map();
        this.scene = scene_1.Scene.create(ctx, 'blended');
        this.parent = parent;
        this._props = { ...param_definition_1.ParamDefinition.getDefaultValues(exports.ImageHelperParams), ...props };
    }
    update() {
        const previousIds = new Set(this.renderObjects.keys());
        this.parent.forEach((r, ro) => {
            if (!ro.state.visible)
                return;
            if (ro.type !== 'image')
                return;
            const values = ro.values;
            const version = values.aPosition.ref.version
                + values.uTrimType.ref.version + values.uTrimCenter.ref.version
                + values.uTrimRotation.ref.version + values.uTrimScale.ref.version
                + values.uTrimTransform.ref.version + values.aTransform.ref.version;
            const existing = this.renderObjects.get(ro.id);
            if (existing && existing.version === version) {
                previousIds.delete(ro.id);
                return;
            }
            // Remove old entries if version changed
            if (existing) {
                for (const oldRO of existing.roList)
                    this.scene.remove(oldRO);
                this.renderObjects.delete(ro.id);
            }
            const roList = [];
            const edgeLines = createImageEdgeLines(values);
            if (edgeLines) {
                const edgeRO = createLinesRenderObject(edgeLines, imageEdgeMaterialId, names_1.ColorNames.cyan, 0.8);
                this.scene.add(edgeRO);
                roList.push(edgeRO);
            }
            const trimLines = createTrimEdgeLines(values);
            if (trimLines) {
                const trimRO = createLinesRenderObject(trimLines, imageTrimMaterialId, names_1.ColorNames.yellow, 0.7);
                this.scene.add(trimRO);
                roList.push(trimRO);
            }
            if (roList.length > 0) {
                this.renderObjects.set(ro.id, { roList, version });
            }
            previousIds.delete(ro.id);
        });
        for (const id of previousIds) {
            const entry = this.renderObjects.get(id);
            if (entry) {
                for (const ro of entry.roList)
                    this.scene.remove(ro);
                this.renderObjects.delete(id);
            }
        }
        this.scene.update(void 0, false);
        this.scene.commit();
    }
    syncVisibility() {
        const visible = this._props.imageEdges;
        this.renderObjects.forEach(entry => {
            for (const ro of entry.roList)
                ro.state.visible = visible;
        });
    }
    clear() {
        this.renderObjects.clear();
        this.scene.clear();
    }
    get isEnabled() {
        return this._props.imageEdges;
    }
    get props() { return this._props; }
    setProps(props) {
        Object.assign(this._props, props);
        if (this.isEnabled)
            this.update();
    }
}
exports.ImageHelper = ImageHelper;
//
/**
 * Image quad vertex layout (from image.ts):
 *   Vertex 0: UV (0,1) — top-left
 *   Vertex 1: UV (0,0) — bottom-left
 *   Vertex 2: UV (1,1) — top-right
 *   Vertex 3: UV (1,0) — bottom-right
 *
 * addPlane expects corners in winding order (0→1→2→3→0),
 * so we reorder to: top-left, bottom-left, bottom-right, top-right.
 */
const _planeCorners = new Float32Array(12);
function createImageEdgeLines(values) {
    const positions = values.aPosition.ref.value;
    const transforms = values.aTransform.ref.value;
    const instanceCount = values.uInstanceCount.ref.value;
    if (positions.length < 12)
        return undefined; // need 4 vertices × 3 components
    // Reorder from [TL, BL, TR, BR] to winding order [TL, BL, BR, TR]
    // V0 (TL) → slot 0
    _planeCorners[0] = positions[0];
    _planeCorners[1] = positions[1];
    _planeCorners[2] = positions[2];
    // V1 (BL) → slot 1
    _planeCorners[3] = positions[3];
    _planeCorners[4] = positions[4];
    _planeCorners[5] = positions[5];
    // V3 (BR) → slot 2
    _planeCorners[6] = positions[9];
    _planeCorners[7] = positions[10];
    _planeCorners[8] = positions[11];
    // V2 (TR) → slot 3
    _planeCorners[9] = positions[6];
    _planeCorners[10] = positions[7];
    _planeCorners[11] = positions[8];
    const builder = lines_builder_1.LinesBuilder.create(4 * instanceCount);
    for (let inst = 0; inst < instanceCount; ++inst) {
        const transform = (0, mat4_1.Mat4)();
        mat4_1.Mat4.fromArray(transform, transforms, inst * 16);
        (0, plane_1.addPlane)(builder, _planeCorners, transform, 0);
    }
    return builder.getLines();
}
function createTrimEdgeLines(values) {
    const trimType = values.uTrimType.ref.value;
    if (trimType === 0)
        return undefined; // no trim
    const transforms = values.aTransform.ref.value;
    const instanceCount = values.uInstanceCount.ref.value;
    vec3_1.Vec3.copy(_trimPos, values.uTrimCenter.ref.value);
    quat_1.Quat.copy(_trimRot, values.uTrimRotation.ref.value);
    vec3_1.Vec3.copy(_trimScale, values.uTrimScale.ref.value);
    mat4_1.Mat4.copy(_trimTransform, values.uTrimTransform.ref.value);
    if (trimType === clip_1.Clip.Type.cube) {
        return createCubeTrimLines(transforms, instanceCount);
    }
    else if (trimType === clip_1.Clip.Type.sphere) {
        return createSphereTrimLines(transforms, instanceCount);
    }
    // For other trim types (plane, cylinder, cone), draw a cube outline as a fallback
    // using the trim center/scale/rotation
    return createCubeTrimLines(transforms, instanceCount);
}
function createCubeTrimLines(transforms, instanceCount) {
    // Build cube transform: translate * rotate * scale
    const rotMat = mat4_1.Mat4.fromQuat((0, mat4_1.Mat4)(), _trimRot);
    const translateMat = mat4_1.Mat4.fromTranslation((0, mat4_1.Mat4)(), _trimPos);
    const scaleMat = mat4_1.Mat4.fromScaling((0, mat4_1.Mat4)(), _trimScale);
    mat4_1.Mat4.mul(_tmpMat, translateMat, rotMat);
    mat4_1.Mat4.mul(_tmpMat, _tmpMat, scaleMat);
    // Apply inverse of trim transform
    if (!mat4_1.Mat4.isIdentity(_trimTransform)) {
        const invTrimTransform = mat4_1.Mat4.invert((0, mat4_1.Mat4)(), _trimTransform);
        mat4_1.Mat4.mul(_tmpMat, invTrimTransform, _tmpMat);
    }
    // addBox uses [0,1]^3, trim cube uses [-0.5,0.5]^3 — prepend offset
    const offset = mat4_1.Mat4.fromTranslation((0, mat4_1.Mat4)(), vec3_1.Vec3.create(-0.5, -0.5, -0.5));
    mat4_1.Mat4.mul(_tmpMat, _tmpMat, offset);
    const builder = lines_builder_1.LinesBuilder.create(12 * instanceCount);
    for (let inst = 0; inst < instanceCount; ++inst) {
        const instTransform = (0, mat4_1.Mat4)();
        mat4_1.Mat4.fromArray(instTransform, transforms, inst * 16);
        const combined = mat4_1.Mat4.mul((0, mat4_1.Mat4)(), instTransform, _tmpMat);
        (0, box_1.addBox)(builder, combined, 0);
    }
    return builder.getLines();
}
function createSphereTrimLines(transforms, instanceCount) {
    const radius = Math.max(_trimScale[0] * 0.5, _trimScale[1] * 0.5, _trimScale[2] * 0.5);
    const rotMat = mat4_1.Mat4.fromQuat((0, mat4_1.Mat4)(), _trimRot);
    const translateMat = mat4_1.Mat4.fromTranslation((0, mat4_1.Mat4)(), _trimPos);
    mat4_1.Mat4.mul(_tmpMat, translateMat, rotMat);
    if (!mat4_1.Mat4.isIdentity(_trimTransform)) {
        const invTrimTransform = mat4_1.Mat4.invert((0, mat4_1.Mat4)(), _trimTransform);
        mat4_1.Mat4.mul(_tmpMat, invTrimTransform, _tmpMat);
    }
    const segments = 32;
    const circlesPerDimension = 3;
    const builder = lines_builder_1.LinesBuilder.create(segments * 3 * circlesPerDimension * instanceCount);
    for (let inst = 0; inst < instanceCount; ++inst) {
        const instTransform = (0, mat4_1.Mat4)();
        mat4_1.Mat4.fromArray(instTransform, transforms, inst * 16);
        const combined = mat4_1.Mat4.mul((0, mat4_1.Mat4)(), instTransform, _tmpMat);
        (0, sphere_1.addSphere)(builder, radius, combined, 0, { segments, circlesPerDimension });
    }
    return builder.getLines();
}
function createLinesRenderObject(lines, materialId, color, alpha) {
    const props = { ...param_definition_1.ParamDefinition.getDefaultValues(lines_1.Lines.Params), sizeFactor: 1, alpha };
    const values = lines_1.Lines.Utils.createValuesSimple(lines, props, color, 1);
    const state = lines_1.Lines.Utils.createRenderableState(props);
    state.pickable = false;
    return (0, render_object_1.createRenderObject)('lines', values, state, materialId);
}
