// molstar/lib/commonjs/extensions/debug-helpers/index.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DebugHelpers = void 0;
const behavior_1 = require("../../mol-plugin/behavior/behavior.js");
const bounding_sphere_helper_1 = require("./bounding-sphere-helper.js");
const clip_object_helper_1 = require("./clip-object-helper.js");
const direct_volume_helper_1 = require("./direct-volume-helper.js");
const image_helper_1 = require("./image-helper.js");
const mesh_helper_1 = require("./mesh-helper.js");
const DebugHelpersParams = {
    ...bounding_sphere_helper_1.BoundingSphereHelperParams,
    ...clip_object_helper_1.ClipObjectHelperParams,
    ...mesh_helper_1.MeshHelperParams,
    ...image_helper_1.ImageHelperParams,
    ...direct_volume_helper_1.DirectVolumeHelperParams,
};
exports.DebugHelpers = behavior_1.PluginBehavior.create({
    name: 'extension-debug-helpers',
    category: 'misc',
    display: {
        name: 'Debug Helpers'
    },
    ctor: class extends behavior_1.PluginBehavior.Handler {
        async register() {
            await this.ctx.canvas3dInitialized;
            const canvas3d = this.ctx.canvas3d;
            if (!canvas3d)
                return;
            const dr = canvas3d.debugRegistry;
            const { ctx, parent } = dr;
            dr.register('bounding-sphere', new bounding_sphere_helper_1.BoundingSphereHelper(ctx, parent, this.params));
            dr.register('clip-object', new clip_object_helper_1.ClipObjectHelper(ctx, parent, this.params));
            dr.register('mesh', new mesh_helper_1.MeshHelper(ctx, parent, this.params));
            dr.register('image', new image_helper_1.ImageHelper(ctx, parent, this.params));
            dr.register('direct-volume', new direct_volume_helper_1.DirectVolumeHelper(ctx, parent, this.params));
        }
        update(params) {
            const changed = super.update(params);
            const canvas3d = this.ctx.canvas3d;
            if (changed && canvas3d) {
                canvas3d.debugRegistry.setProps(params);
                canvas3d.requestDraw();
            }
            return changed;
        }
        unregister() {
            const canvas3d = this.ctx.canvas3d;
            if (!canvas3d)
                return;
            const dr = canvas3d.debugRegistry;
            dr.unregister('bounding-sphere');
            dr.unregister('clip-object');
            dr.unregister('mesh');
            dr.unregister('image');
            dr.unregister('direct-volume');
        }
    },
    params: () => DebugHelpersParams,
    canAutoUpdate: () => true,
});
