// molstar/lib/commonjs/extensions/debug-helpers/mesh-helper.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeshHelper = exports.MeshHelperParams = void 0;
const render_object_1 = require("../../mol-gl/render-object.js");
const param_definition_1 = require("../../mol-util/param-definition.js");
const scene_1 = require("../../mol-gl/scene.js");
const names_1 = require("../../mol-util/color/names.js");
const lines_1 = require("../../mol-geo/geometry/lines/lines.js");
const lines_builder_1 = require("../../mol-geo/geometry/lines/lines-builder.js");
const vec3_1 = require("../../mol-math/linear-algebra/3d/vec3.js");
const mat4_1 = require("../../mol-math/linear-algebra/3d/mat4.js");
exports.MeshHelperParams = {
    meshNormals: param_definition_1.ParamDefinition.Boolean(false, { description: 'Show normals of visible mesh render objects.' }),
};
const meshHelperMaterialId = (0, render_object_1.getNextMaterialId)();
const _v = (0, vec3_1.Vec3)();
const _n = (0, vec3_1.Vec3)();
const _start = (0, vec3_1.Vec3)();
const _end = (0, vec3_1.Vec3)();
class MeshHelper {
    constructor(ctx, parent, props) {
        this.renderObjects = new Map();
        this.scene = scene_1.Scene.create(ctx, 'blended');
        this.parent = parent;
        this._props = { ...param_definition_1.ParamDefinition.getDefaultValues(exports.MeshHelperParams), ...props };
    }
    update() {
        const previousIds = new Set(this.renderObjects.keys());
        const currentIds = new Set();
        this.parent.forEach((r, ro) => {
            if (!ro.state.visible)
                return;
            if (ro.type !== 'mesh')
                return;
            currentIds.add(ro.id);
            // Skip if we already have normals for this render object
            if (this.renderObjects.has(ro.id)) {
                previousIds.delete(ro.id);
                return;
            }
            const values = ro.values;
            const lines = createNormalLines(values);
            if (!lines)
                return;
            const linesRO = createNormalLinesRenderObject(lines, meshHelperMaterialId);
            this.scene.add(linesRO);
            this.renderObjects.set(ro.id, linesRO);
        });
        // Remove normals for render objects no longer present
        for (const id of previousIds) {
            const linesRO = this.renderObjects.get(id);
            if (linesRO) {
                this.scene.remove(linesRO);
                this.renderObjects.delete(id);
            }
        }
        this.scene.update(void 0, false);
        this.scene.commit();
    }
    syncVisibility() {
        const visible = this._props.meshNormals;
        this.renderObjects.forEach(ro => {
            ro.state.visible = visible;
        });
    }
    clear() {
        this.renderObjects.clear();
        this.scene.clear();
    }
    get isEnabled() {
        return this._props.meshNormals;
    }
    get props() { return this._props; }
    setProps(props) {
        Object.assign(this._props, props);
        if (this.isEnabled)
            this.update();
    }
}
exports.MeshHelper = MeshHelper;
//
function createNormalLines(values) {
    const positions = values.aPosition.ref.value;
    const normals = values.aNormal.ref.value;
    const indices = values.elements.ref.value;
    const transforms = values.aTransform.ref.value;
    const instanceCount = values.uInstanceCount.ref.value;
    const vertexCount = positions.length / 3;
    if (vertexCount === 0)
        return undefined;
    // Determine normal line length: proportional to bounding sphere radius
    const bs = values.boundingSphere.ref.value;
    const normalLength = Math.max(bs.radius * 0.01, 0.1);
    // Count unique vertices referenced by indices
    const indexCount = values.drawCount.ref.value;
    const builder = lines_builder_1.LinesBuilder.create(indexCount * instanceCount);
    for (let inst = 0; inst < instanceCount; ++inst) {
        const tOffset = inst * 16;
        const transform = (0, mat4_1.Mat4)();
        mat4_1.Mat4.fromArray(transform, transforms, tOffset);
        // Use a set to avoid drawing duplicate normals for shared vertices
        const visited = new Set();
        for (let i = 0; i < indexCount; ++i) {
            const vi = indices[i];
            if (visited.has(vi))
                continue;
            visited.add(vi);
            const vo = vi * 3;
            vec3_1.Vec3.set(_v, positions[vo], positions[vo + 1], positions[vo + 2]);
            vec3_1.Vec3.set(_n, normals[vo], normals[vo + 1], normals[vo + 2]);
            // Transform vertex position and normal direction by instance transform
            vec3_1.Vec3.transformMat4(_start, _v, transform);
            vec3_1.Vec3.transformDirection(_end, _n, transform);
            vec3_1.Vec3.normalize(_end, _end);
            vec3_1.Vec3.scaleAndAdd(_end, _start, _end, normalLength);
            builder.addVec(_start, _end, 0);
        }
    }
    return builder.getLines();
}
function createNormalLinesRenderObject(lines, materialId) {
    const props = { ...param_definition_1.ParamDefinition.getDefaultValues(lines_1.Lines.Params), sizeFactor: 1, alpha: 0.7 };
    const values = lines_1.Lines.Utils.createValuesSimple(lines, props, names_1.ColorNames.magenta, 1);
    const state = lines_1.Lines.Utils.createRenderableState(props);
    state.pickable = false;
    return (0, render_object_1.createRenderObject)('lines', values, state, materialId);
}
