// molstar/lib/commonjs/mol-model/structure/structure/coordination/compute.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeCoordination = computeCoordination;
const unit_1 = require("../unit.js");
const element_1 = require("../element.js");
const data_1 = require("./data.js");
const util_1 = require("../../../../mol-data/util.js");
const types_1 = require("../../model/types.js");
const measures_1 = require("../../model/properties/atomic/measures.js");
/** Minimum number of bonds for an atom to be considered a coordination site */
const MinCoordination = 4;
function interBondCount(structure, unit, index) {
    let count = 0;
    const indices = structure.interUnitBonds.getEdgeIndices(index, unit.id);
    for (let i = 0, il = indices.length; i < il; ++i) {
        const b = structure.interUnitBonds.edges[indices[i]];
        if (types_1.BondType.isCovalent(b.props.flag) || types_1.BondType.is(b.props.flag, types_1.BondType.Flag.MetallicCoordination))
            count += 1;
    }
    return count;
}
function intraBondCount(unit, index) {
    let count = 0;
    const { offset, edgeProps: { flags } } = unit.bonds;
    for (let i = offset[index], il = offset[index + 1]; i < il; ++i) {
        if (types_1.BondType.isCovalent(flags[i]) || types_1.BondType.is(flags[i], types_1.BondType.Flag.MetallicCoordination))
            count += 1;
    }
    return count;
}
function bondCount(structure, unit, index) {
    return interBondCount(structure, unit, index) + intraBondCount(unit, index);
}
//
const _H = measures_1.AtomicNumbers['H'];
function eachInterBondedAtom(structure, unit, index, cb) {
    const indices = structure.interUnitBonds.getEdgeIndices(index, unit.id);
    for (let i = 0, il = indices.length; i < il; ++i) {
        const b = structure.interUnitBonds.edges[indices[i]];
        const uB = structure.unitMap.get(b.unitB);
        if ((types_1.BondType.isCovalent(b.props.flag) || types_1.BondType.is(b.props.flag, types_1.BondType.Flag.MetallicCoordination)) && uB.model.atomicHierarchy.derived.atom.atomicNumber[uB.elements[b.indexB]] !== _H)
            cb(uB, b.indexB);
    }
}
function eachIntraBondedAtom(unit, index, cb) {
    const { elements } = unit;
    const { atomicNumber } = unit.model.atomicHierarchy.derived.atom;
    const { offset, b, edgeProps: { flags } } = unit.bonds;
    for (let i = offset[index], il = offset[index + 1]; i < il; ++i) {
        if ((types_1.BondType.isCovalent(flags[i]) || types_1.BondType.is(flags[i], types_1.BondType.Flag.MetallicCoordination)) && atomicNumber[elements[b[i]]] !== _H)
            cb(unit, b[i]);
    }
}
function eachBondedAtom(structure, unit, index, cb) {
    eachInterBondedAtom(structure, unit, index, cb);
    eachIntraBondedAtom(unit, index, cb);
}
//
function computeCoordination(structure) {
    const unitIds = [];
    const indices = [];
    const numbers = [];
    const siteIndex = new Map();
    for (let ui = 0, uil = structure.units.length; ui < uil; ++ui) {
        const unit = structure.units[ui];
        if (!unit_1.Unit.isAtomic(unit))
            continue;
        const { elements } = unit;
        for (let ei = 0, eil = elements.length; ei < eil; ++ei) {
            const element = elements[ei];
            const unitIndex = ei;
            const _bondCount = bondCount(structure, unit, unitIndex);
            if (_bondCount >= MinCoordination) {
                siteIndex.set(coordinationKey(unit.id, element), siteIndex.size);
                unitIds.push(unit.id);
                indices.push(unitIndex);
                numbers.push(_bondCount);
            }
        }
    }
    if (siteIndex.size === 0)
        return data_1.EmptyCoordination;
    const l = element_1.StructureElement.Location.create(structure);
    return {
        sites: {
            unitIds,
            indices,
            numbers,
            count: siteIndex.size
        },
        getSiteIndex: (unit, element) => {
            var _a;
            return (_a = siteIndex.get(coordinationKey(unit.id, element))) !== null && _a !== void 0 ? _a : -1;
        },
        eachLigand: (siteIndex, cb) => {
            const unitId = unitIds[siteIndex];
            const element = indices[siteIndex];
            const unit = structure.unitMap.get(unitId);
            eachBondedAtom(structure, unit, element, (u, i) => {
                l.unit = u;
                l.element = u.elements[i];
                cb(l);
            });
        }
    };
}
function coordinationKey(unitId, element) {
    return (0, util_1.cantorPairing)(unitId, element);
}
