// molstar/lib/mol-canvas3d/helper/debug-registry.js
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
import { isDebugMode } from '../../mol-util/debug.js';
export class DebugRegistry {
    constructor(ctx, parent) {
        this.entries = new Map();
        this.ctx = ctx;
        this.parent = parent;
    }
    register(name, entry) {
        if (this.entries.has(name)) {
            if (isDebugMode) {
                console.warn(`Debug helper with name '${name}' already exists, replacing.`);
            }
            this.entries.get(name).clear();
        }
        this.entries.set(name, entry);
    }
    unregister(name) {
        const entry = this.entries.get(name);
        if (entry) {
            entry.clear();
            this.entries.delete(name);
        }
    }
    get scenes() {
        return Array.from(this.entries.values()).map(e => e.scene);
    }
    update() {
        this.entries.forEach(entry => {
            if (entry.isEnabled)
                entry.update();
        });
    }
    syncVisibility() {
        this.entries.forEach(entry => {
            entry.syncVisibility();
        });
    }
    clear() {
        this.entries.forEach(entry => {
            entry.clear();
        });
    }
    get isEnabled() {
        let enabled = false;
        this.entries.forEach(entry => {
            if (entry.isEnabled)
                enabled = true;
        });
        return enabled;
    }
    setProps(props) {
        this.entries.forEach(entry => {
            entry.setProps(props);
        });
    }
}
