// molstar/lib/commonjs/apps/viewer/extensions.js
"use strict";
/**
 * Copyright (c) 2018-2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginExtensions = exports.ExtensionMap = void 0;
const tslib_1 = require("tslib");
const behavior_1 = require("../../extensions/anvil/behavior.js");
const assembly_symmetry_1 = require("../../extensions/assembly-symmetry/index.js");
const backgrounds_1 = require("../../extensions/backgrounds/index.js");
const debug_helpers_1 = require("../../extensions/debug-helpers/index.js");
const dnatco_1 = require("../../extensions/dnatco/index.js");
const format_1 = require("../../extensions/g3d/format.js");
const geo_export_1 = require("../../extensions/geo-export/index.js");
const behavior_2 = require("../../extensions/model-archive/quality-assessment/behavior.js");
const model_export_1 = require("../../extensions/model-export/index.js");
const mp4_export_1 = require("../../extensions/mp4-export/index.js");
const mvs_1 = require("../../extensions/mvs/index.js");
const behavior_3 = require("../../extensions/mvs/behavior.js");
const formats_1 = require("../../extensions/mvs/components/formats.js");
const pdbe_1 = require("../../extensions/pdbe/index.js");
const rcsb_1 = require("../../extensions/rcsb/index.js");
const sb_ncbr_1 = require("../../extensions/sb-ncbr/index.js");
const behavior_4 = require("../../extensions/wwpdb/ccd/behavior.js");
const struct_conn_1 = require("../../extensions/wwpdb/struct-conn/index.js");
const zenodo_1 = require("../../extensions/zenodo/index.js");
const spec_1 = require("../../mol-plugin/spec.js");
const mvs_data_1 = require("../../extensions/mvs/mvs-data.js");
const MVSUtil = tslib_1.__importStar(require("../../extensions/mvs/util.js"));
exports.ExtensionMap = {
    // Mol* built-in extensions
    'mvs': spec_1.PluginSpec.Behavior(behavior_3.MolViewSpec),
    'backgrounds': spec_1.PluginSpec.Behavior(backgrounds_1.Backgrounds),
    'debug-helpers': spec_1.PluginSpec.Behavior(debug_helpers_1.DebugHelpers),
    'model-export': spec_1.PluginSpec.Behavior(model_export_1.ModelExport),
    'mp4-export': spec_1.PluginSpec.Behavior(mp4_export_1.Mp4Export),
    'geo-export': spec_1.PluginSpec.Behavior(geo_export_1.GeometryExport),
    'zenodo-import': spec_1.PluginSpec.Behavior(zenodo_1.ZenodoImport),
    'wwpdb-chemical-component-dictionary': spec_1.PluginSpec.Behavior(behavior_4.wwPDBChemicalComponentDictionary),
    // 3rd party extensions
    'pdbe-structure-quality-report': spec_1.PluginSpec.Behavior(pdbe_1.PDBeStructureQualityReport),
    'dnatco-ntcs': spec_1.PluginSpec.Behavior(dnatco_1.DnatcoNtCs),
    'assembly-symmetry': spec_1.PluginSpec.Behavior(assembly_symmetry_1.AssemblySymmetry),
    'rcsb-validation-report': spec_1.PluginSpec.Behavior(rcsb_1.RCSBValidationReport),
    'anvil-membrane-orientation': spec_1.PluginSpec.Behavior(behavior_1.ANVILMembraneOrientation),
    'g3d': spec_1.PluginSpec.Behavior(format_1.G3DFormat), // TODO: consider removing this for Mol* 6.0
    'ma-quality-assessment': spec_1.PluginSpec.Behavior(behavior_2.MAQualityAssessment),
    'sb-ncbr-partial-charges': spec_1.PluginSpec.Behavior(sb_ncbr_1.SbNcbrPartialCharges),
    'tunnels': spec_1.PluginSpec.Behavior(sb_ncbr_1.SbNcbrTunnels),
};
exports.PluginExtensions = {
    wwPDBStructConn: struct_conn_1.wwPDBStructConnExtensionFunctions,
    mvs: {
        MVSData: mvs_data_1.MVSData,
        createBuilder: mvs_data_1.MVSData.createBuilder,
        loadMVS: mvs_1.loadMVS,
        loadMVSData: formats_1.loadMVSData,
        util: {
            ...MVSUtil
        }
    },
    modelArchive: {
        qualityAssessment: {
            config: behavior_2.MAQualityAssessmentConfig
        }
    }
};
